File: src/Simbuino.UI/Audio/AudioPlayer.cs
using SharpDX;
using SharpDX.Multimedia;
using SharpDX.XAudio2;
using Simbuino.Emulator;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace Simbuino.UI.Audio
{
	public static class AudioPlayer
	{
		public const int AudioFrequency = 44100;
		public const int AudioSlices = 20; // 1 audio block per default frame 
		public const int SamplesPerSlice = AudioFrequency / AudioSlices;
		public const int CyclesPerSample = AtmelProcessor.ClockSpeed / AudioFrequency;

		private static AudioBuffer[] AudioBuffers = new AudioBuffer[2];
		private static AutoResetEvent[] BufferEvents = new AutoResetEvent[2];
		private static byte[][] SampleBuffers = new byte[2][];
		private static int CurrentBuffer = 0;
		private static int CurrentSample = 0;
		private static int LastSample = 0;
		private static int FilteredSample = 0;
		private static int FilterValue = 0;
		private static long LastCycle = 0;
		private static SourceVoice SourceVoice;
		
		private static bool _Filtered = false;
		public static bool Filtered
		{
			get { return _Filtered; }
			set
			{
				_Filtered = value;
				FilterValue = value ? 220 : 0;
			}
		}

		public volatile static bool Playing = false;


		static AudioPlayer()
		{
			SampleBuffers[0] = new byte[AudioFrequency / AudioSlices];			
			var dataStream0 = DataStream.Create(SampleBuffers[0], true, true);
			AudioBuffers[0] = new AudioBuffer
			{
				Stream = dataStream0,
				AudioBytes = (int)dataStream0.Length,
				Flags = BufferFlags.None,
				Context = new IntPtr(0)
			};
			BufferEvents[0] = new AutoResetEvent(false);

			SampleBuffers[1] = new byte[AudioFrequency / AudioSlices];
			var dataStream1 = DataStream.Create(SampleBuffers[1], true, true);
			AudioBuffers[1] = new AudioBuffer
			{
				Stream = dataStream1,
				AudioBytes = (int)dataStream1.Length,
				Flags = BufferFlags.None,
				Context = new IntPtr(1)
			};
			BufferEvents[1] = new AutoResetEvent(false);

			var waveFormat = new WaveFormat(AudioFrequency, 8, 1);
			XAudio2 xaudio = new XAudio2();
			MasteringVoice masteringVoice = new MasteringVoice(xaudio);
			SourceVoice = new SourceVoice(xaudio, waveFormat, true);
			SourceVoice.BufferStart += SourceVoice_BufferStart;
		}

		public static void Start()
		{			
			CurrentBuffer = 0;

			// queue 2 empty buffers, we'll start filling the second one
			SourceVoice.Start();
			QueueBuffer();
			Playing = true;

			var reg = AtmelContext.RAM[AtmelIO.OCR2B] as ObservableRegister;
			reg.OnRegisterChanged += reg_OnRegisterChanged;
		}

		private static void QueueBuffer()
		{
			// fill remaining buffer with the current sample
			var buffer = SampleBuffers[CurrentBuffer];
			while (CurrentSample < SamplesPerSlice)
			{
				FilteredSample = (FilteredSample * FilterValue + LastSample * (256 - FilterValue)) / 256;
				buffer[CurrentSample++] = (byte)FilteredSample;
			}

			// queue the buffer and wait for it to start playing			
			BufferEvents[CurrentBuffer].Reset();
			SourceVoice.SubmitSourceBuffer(AudioBuffers[CurrentBuffer], null);
			BufferEvents[CurrentBuffer].WaitOne();
			CurrentBuffer = 1 - CurrentBuffer;
			CurrentSample = 0;
			LastCycle = AtmelContext.Clock;
		}

		public static void Stop()
		{
			Playing = false;
			var reg = AtmelContext.RAM[AtmelIO.OCR2B] as ObservableRegister;
			reg.OnRegisterChanged -= reg_OnRegisterChanged;
			SourceVoice.Stop();
		}

		static void reg_OnRegisterChanged(int oldVal, int newVal)
		{
			if (!Playing)
				return;

			// if the buffer is full then we're going to have to skip samples
			if (CurrentSample >= SamplesPerSlice)
				return;

			// which sample are we on?
			int sampleNum = Math.Min(SamplesPerSlice, (int)((AtmelContext.Clock - LastCycle) / CyclesPerSample));

			// fill the buffer up to this point with the old value
			var buffer = SampleBuffers[CurrentBuffer];
			while (CurrentSample < sampleNum)
			{
				FilteredSample = (FilteredSample * FilterValue + LastSample * (256 - FilterValue)) / 256;
				buffer[CurrentSample++] = (byte)FilteredSample;
			}
			CurrentSample = sampleNum;

			// make note of our new value
			if (LastSample != newVal)
				LastSample = (byte)newVal;
		}

		public static void NextBuffer()
		{
			QueueBuffer();
		}

		private static void SourceVoice_BufferStart(IntPtr obj)
		{
			BufferEvents[obj.ToInt32()].Set();
		}

	}
}

File: binaries/Simbuino4Web/Scripts/Views/AudioPlayer.js
$(function () {
	AudioPlayer = Class.create({

		ctor: function () {
			var self = this;
			this.BufferSize = 4096;
			this.StartCycle = AtmelContext.Clock;
			this.CurrentSample = 0;
			this.StartCycle = 10;
			this.LastSample = 0;
			var AudioContext = window.AudioContext || window.webkitAudioContext;
			if (AudioContext)
			{
				this.Context = new AudioContext();
				this.SampleRate = this.Context.sampleRate;				
				
				this.CyclesPerSample = Math.floor(AtmelProcessor.ClockSpeed / this.SampleRate);
				this.MixBuffer = [];
				for (var i = 0; i < this.BufferSize; i++)
					this.MixBuffer[i] = 0;
				this.Processor = this.Context.createScriptProcessor(this.BufferSize, 1, 1);
				this.Processor.onaudioprocess = function (e) {

					var src = self.MixBuffer;
					var dst = e.outputBuffer.getChannelData(0);
					var count = Math.min(self.CurrentSample, self.BufferSize);

					// copy over samples we've done for this buffer
					for (var i = 0; i < count; i++)
						dst[i] = src[i];

					// fill any remaining samples with the last sample value
					for (var i = count; i < self.BufferSize; i++)
						dst[i] = self.LastSample;

					self.StartCycle = AtmelContext.Clock;
					self.CurrentSample = 0;
				};

				// intercept writes to the OCR2B sound register and start playing
				AtmelContext.RAM[AtmelIO.OCR2B].OnRegisterChanged.push(function (oldVal, newVal) { self.reg_OnRegisterChanged(oldVal, newVal) });
				this.Processor.connect(this.Context.destination);
				this.Active = true;
			}
		},

		reg_OnRegisterChanged: function(oldVal, newVal)
		{
			newVal = (newVal - 128) / 512;

			// which sample are we on?
			var sampleNum = Math.floor((AtmelContext.Clock - this.StartCycle) / this.CyclesPerSample);

			// make sure we don't get too far ahead				
			sampleNum = Math.min(sampleNum, this.BufferSize-1);

			// fill the buffer up to this point with the old value
			var buffer = this.MixBuffer;
			while (this.CurrentSample < sampleNum)
				buffer[this.CurrentSample++] = newVal;

			// make note of our new value
			this.LastSample = newVal;
		}

	});

});
